### PR TITLE
Item processing stability

### PR DESCRIPTION
**Pull request by Ziktofel**
## What is this fixing or adding?
See #309 and #310

Adds ItemFilterFlags to be processable by `pool_filter`
Start of splitting the removal of hero items and the related faction items as a preparation for cases where the player will want to play with Nova + Zerg, Kerrigan + Terran, etc
W/A upgrades aren't any longer locked by mission count (they are part of the logic instead)
## How was this tested?
Generation and unit tests

## If this makes graphical changes, please attach screenshots.
No

### CODE REVIEW

File: worlds/sc2/item/item_groups.py [hunk 1]
```diff
@@ -424,8 +425,9 @@ def get_all_group_names(cls) -> typing.Set[str]:
     item_names.NYDUS_WORM,
     item_names.OMEGA_WORM]
 item_name_groups[ItemGroupNames.ZERG_UNITS] = zerg_units = [
-    item_name for item_name, item_data in items.item_table.items()
-    if item_data.type in (items.ZergItemType.Unit, items.ZergItemType.Mercenary, items.ZergItemType.Morph)
+    item_name for item_name, item_data in item_tables.item_table.items()
+    if item_data.type in (
+    item_tables.ZergItemType.Unit, item_tables.ZergItemType.Mercenary, item_tables.ZergItemType.Morph)
```
Comment (MatthewMarinets) [inline worlds/sc2/item/item_groups.py]: (minor) looks like a missing indent here in in embedded tuple.

File: worlds/sc2/pool_filter.py [hunk 2]
```diff
@@ -90,6 +92,29 @@ def attempt_removal(item: Item) -> bool:
                     return False
             return True
 
+        # Process Excluded items, validate if the item can get actually excluded
+        excluded_items: List[StarcraftItem] = [stacrcraft_item for stacrcraft_item in inventory if ItemFilterFlags.Excluded in stacrcraft_item.filter_flags]
```
Comment (MatthewMarinets) [inline worlds/sc2/pool_filter.py]: (minor) typo in variable name `stacrcraft_item` -> `starcraft_item`

File: worlds/sc2/pool_filter.py [hunk 3]
```diff
@@ -157,16 +182,16 @@ def attempt_removal(item: Item) -> bool:
             raise Exception(f"Too many items excluded - couldn't satisfy access rules for the following locations:\n{failed_locations}")
 
         # Optionally locking generic items
-        generic_items = [item for item in inventory if item.name in second_pass_placeable_items]
+        generic_items: List[StarcraftItem] = [stacrcraft_item for stacrcraft_item in inventory if stacrcraft_item.name in second_pass_placeable_items]
```
Comment (MatthewMarinets) [inline worlds/sc2/pool_filter.py]: Same typo - stacrcraft item. Here (line 185) and below (line 190~194)

File: worlds/sc2/item/__init__.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    index: int = 0
+    flags: ItemFilterFlags = ItemFilterFlags.Available
+
+
+class StarcraftItem(Item):
+    game: str = "Starcraft 2"
+    filter_flags: ItemFilterFlags = ItemFilterFlags.Available
+
+    def __init__(self, name: str, classification: ItemClassification, code: Optional[int], player: int, filter_flags: ItemFilterFlags = ItemFilterFlags.Available):
+        Item.__init__(self, name, classification, code, player)
```
Comment (MatthewMarinets) [inline worlds/sc2/item/__init__.py]: Usually I see `super().__init__()` (or some variation, I do OOP infrequently enough I have to look it up every time).

Also missing a newline at the end of this file.

File: worlds/sc2/__init__.py [hunk 5]
```diff
@@ -377,6 +343,8 @@ def resolve_count(count: Optional[int], max_count: int) -> int:
                 result[-1].flags |= ItemFilterFlags.Locked
```
Comment (MatthewMarinets) [inline worlds/sc2/__init__.py:343]: From what I've seen so far, "locked" and "necessary" are still unchanged (one task suggests renaming them from "locked", "necessary" -> "requested", "locked"). This means the user option is still hooked up to the flag that can be overridden, which isn't what we want, I think. So this should add the "necessary" flag (we can handle the rename later).

File: worlds/sc2/rules.py [hunk 6]
```diff
@@ -148,7 +147,7 @@ def terran_air(self, state: CollectionState) -> bool:
         """
         return (state.has_any({item_names.VIKING, item_names.WRAITH, item_names.BANSHEE, item_names.BATTLECRUISER}, self.player) or self.advanced_tactics
                 and state.has_any({item_names.HERCULES, item_names.MEDIVAC}, self.player) and self.terran_common_unit(state)
-        )
+                )
```
Comment (MatthewMarinets) [inline worlds/sc2/rules.py:156]: (minor) I don't think this indentation was desirable, a weird PyCharm default?

File: worlds/sc2/test/test_generation.py [hunk 7]
```diff
@@ -779,3 +780,39 @@ def test_weapon_armor_upgrades_generic_upgrade_missions_no_countermeasure_needed
...
+            'custom_mission_order': {
+                'campaign': {
+                    'goal': True,
+                    'layout': {
+                        'type': 'column',
+                        'size': 2,
+                        'missions': [
+                            {
+                                'index': 0,
+                                'mission_pool': ['Liberation Day']
```
Comment (MatthewMarinets) [inline worlds/sc2/test/test_generation.py]: You can use `mission_tables.SC2Mission.LIBERATION_DAY.mission_name` here (and for Supreme below and Infinite Cycle in test_usecases.py). It's more robust to us changing the name down the line, which may happen with race-swaps. I don't think we're going to raceswap no-builds anytime soon, but if we do this is just a minor headache for whoever implements it. Would also happen if we did some kind of mission variant, like a macro version.

**Review by MatthewMarinets — APPROVED** [review]
Lots of changes, I kind of skimmed over most of the rename stuff and that means I didn't go through the removed/excluded stuff in init.py with a fine-toothed comb. Aside from the user option pointing to the "locked" flag rather than "necessary", things look good.

Overall, this seems to cover most of #309 -- the hard part at least. The remainder is the "allow_generation_failures" option, which we can delay comfortably; and various naming/cleanup things like removing `origin`.

File: worlds/sc2/__init__.py (hunk 5 above)
Comment (Ziktofel) [reply worlds/sc2/__init__.py:343]: This is a preparation step. Mentioned that in #309

**Pull request merged**